components/nostr-auth.js

```diff
@@ -74,83 +89,186 @@ export function NostrAuth ({ text, callbackUrl, multiAuth }) {
     // don't cache this mutation
     fetchPolicy: 'no-cache'
   })
-  const [hasExtension, setHasExtension] = useState(undefined)
-  const [extensionError, setExtensionError] = useState(null)
 
-  useEffect(() => {
-    createAuth()
-    setHasExtension(!!window.nostr)
+  // print an error message
+  const setError = useCallback((e) => {
+    console.error(e)
+    toaster.danger(e.message || e.toString())
+    setStatus({
+      msg: e.message || e.toString(),
+      error: true,
+      loading: false
+    })
   }, [])
 
-  const k1 = data?.createAuth.k1
-
-  useEffect(() => {
-    if (!k1 || !hasExtension) return
-
-    console.info('nostr extension detected')
-
-    let mounted = true;
-    (async function () {
-      try {
-        // have them sign a message with the challenge
-        let event
-        try {
-          event = await callWithTimeout(() => window.nostr.signEvent({
-            kind: 22242,
-            created_at: Math.floor(Date.now() / 1000),
-            tags: [['challenge', k1]],
-            content: 'Stacker News Authentication'
-          }), 5000)
-          if (!event) throw new Error('extension returned empty event')
-        } catch (e) {
-          if (e.message === 'window.nostr call already executing' || !mounted) return
-          setExtensionError({ message: 'nostr extension failed to sign event', details: e.message })
-          return
-        }
-
-        // sign them in
-        try {
-          await signIn('nostr', {
-            event: JSON.stringify(event),
-            callbackUrl,
-            multiAuth
-          })
-        } catch (e) {
-          throw new Error('authorization failed', e)
-        }
-      } catch (e) {
-        if (!mounted) return
-        console.log('nostr auth error', e)
-        setExtensionError({ message: `${text} failed`, details: e.message })
+  // authorize user
+  const auth = useCallback(async (nip46token) => {
```

Comment:
Since NDK has this nice signer abstraction, i've made this single callback to handle both nip46 and nip07 logins.